### PR TITLE
[BOLT] Add reading support for Linux kernel .altinstructions section

### PR DESCRIPTION
**Pull request by maksfb**
Read .altinstructions and annotate instructions that have alternative sequences with "AltInst" annotation. Note that some instructions may have more than one alternatives, in which case they will have multiple annotations in the form "AltInst", "AltInst2", "AltInst3", etc.

### CODE REVIEW

**Comment by maksfb** [comment]
Need to add a test case.

File: bolt/lib/Rewrite/LinuxKernelRewriter.cpp [hunk 1]
```diff
@@ -27,6 +27,21 @@ using namespace bolt;
...
+                     cl::init(false), cl::Hidden, cl::cat(BoltCategory));
+
+static cl::opt<uint32_t>
+    AltInstFeatureSize("alt-inst-feature-size",
+                       cl::desc("size of feature field in .altinstructions"),
+                       cl::init(2), cl::Hidden, cl::cat(BoltCategory));
+
+static cl::opt<bool>
+    DumpAltInstructions("dump-alt-instructions",
+                        cl::desc("dump Linux alernative instructions info"),
```
Comment (dcci) [inline bolt/lib/Rewrite/LinuxKernelRewriter.cpp:42]: typo: alternative.

File: bolt/lib/Rewrite/LinuxKernelRewriter.cpp [hunk 2]
```diff
@@ -1132,6 +1156,123 @@ Error LinuxKernelRewriter::readBugTable() {
...
+
+  const uint64_t Address = AltInstrSection->getAddress();
+  DataExtractor DE = DataExtractor(AltInstrSection->getContents(),
+                                   BC.AsmInfo->isLittleEndian(),
+                                   BC.AsmInfo->getCodePointerSize());
+  uint64_t EntryID = 0;
+  DataExtractor::Cursor Cursor(0);
+  while (Cursor && !DE.eof(Cursor)) {
+    const uint64_t OrgInstAddress =
+        Address + Cursor.tell() + (int32_t)DE.getU32(Cursor);
```
Comment (dcci) [inline bolt/lib/Rewrite/LinuxKernelRewriter.cpp:1188]: here you're parsing a signed 32-bit value as unsigned then casting. Is it always correct?

File: bolt/lib/Rewrite/LinuxKernelRewriter.cpp [hunk 3]
```diff
@@ -1132,6 +1156,123 @@ Error LinuxKernelRewriter::readBugTable() {
...
+/// The kernel can replace certain instruction sequences depending on hardware
+/// it is running on and features specified during boot time. The information
+/// about alternative instruction sequences is stored in .altinstructions
+/// section. The format of entries in this section is defined in
+/// arch/x86/include/asm/alternative.h:
+///
+///   struct alt_instr {
+///     s32 instr_offset;
+///     s32 repl_offset;
+///     uXX feature;
```
Comment (dcci) [inline bolt/lib/Rewrite/LinuxKernelRewriter.cpp:1168]: what does uXX stand for?

File: bolt/lib/Rewrite/LinuxKernelRewriter.cpp (hunk 2 above)
Comment (maksfb) [reply bolt/lib/Rewrite/LinuxKernelRewriter.cpp:1188]: Yes. That's what `DataExtractor::getSigned()` does underneath. I would have used it, but there's no `Cursor` version of this interface.

File: bolt/lib/Rewrite/LinuxKernelRewriter.cpp (hunk 3 above)
Comment (maksfb) [reply bolt/lib/Rewrite/LinuxKernelRewriter.cpp:1168]: u16 or u32. Perhaps u64 in the future.

**Review by dcci — APPROVED** [review]
LGTM. (I already reviewed it -- appreciate you clarifying my comments). Thanks for adding a test.

**Comment by maksfb** [comment]
Thanks for the review.

File: bolt/lib/Rewrite/LinuxKernelRewriter.cpp (hunk 2 above)
Comment (ayermolo) [reply bolt/lib/Rewrite/LinuxKernelRewriter.cpp:1188]: Doesn't coding guide prefers using c++ style casting?
https://llvm.org/docs/CodingStandards.html#prefer-c-style-casts

Comment (maksfb) [reply bolt/lib/Rewrite/LinuxKernelRewriter.cpp:1188]: The rule lists an exception for integral type casts.

**Pull request merged**